app/Connect/Actions/GetLatestClientVersionAction.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Connect\Actions;
+
+use App\Connect\Support\BaseApiAction;
+use App\Models\Emulator;
+use Illuminate\Http\Request;
+
+class GetLatestClientVersionAction extends BaseApiAction
+{
+    protected int $emulatorId;
+
+    public function execute(int $emulatorId): array
+    {
+        $this->emulatorId = $emulatorId;
+
+        return $this->process();
+    }
+
+    protected function initialize(Request $request): ?array
+    {
+        if (!$request->has(['e'])) {
+            return $this->missingParameters();
+        }
+
+        $this->emulatorId = request()->integer('e', 0);
+
+        return null;
+    }
+
+    protected function process(): array
+    {
+        $emulator = Emulator::find($this->emulatorId);
+        if ($emulator === null || !$emulator->active || !$emulator->latestRelease) {
+            return $this->resourceNotFound('client');
+        }
+
+        // turn relative URLs into full URLs
+        $format_url = function (?string $url): ?string {
+            return (!$url || str_starts_with($url, 'http')) ? $url : config('app.url') . '/' . $url;
+        };
```

Comment:
nit: I know it's copied from _dorequest.php_, but I wonder if this may be better-suited as a private function of the class, just for consistency.

The comment could probably be eliminated if the function name better described what the function was doing, eg: `$this->makeAbsoluteUrl()` or `$this->toFullUrl()`.

Reply:
done.